### PR TITLE
wip/6.0 HHH-14607 prefer FetchOptions over FetchStrategy

### PR DESCRIPTION
**Pull request by NathanQingyangXu**
https://hibernate.atlassian.net/browse/HHH-14607

The ultimate goal is to avoid instantiating **FetchStrategy** (currently the only implementation of **FetchOptions** interface which is simply a combination of **FetchTimingAccess** interface and **FetchStyleAccess** interface) to eliminate a performance drain.

In the new v6 mapping stuff, it is easy to achieve that by simply creating various overloaded constructors which accepts **FetchTiming** and **FetchStyle** in instead of **FetchOptions** (or even worse, **FetchStrategy**).

In the deprecated v5 loading stuff (mainly in **walker** package), I went even further to delete **FetchStrategy** totally by:

1. substitute **FetchOptions** for **FetchStrategy** in method signature(s);
2. create a static factory method (i.e. **FetchOptions**.**valueOf**(**FetchTiming** timing, **FetchStyle** style)) and use it to return **FetchOptions** instance (internally a package private **FetchOptionsImpl** class was created but the whole codebase needs to know nothing about its existence outside the package where both **FetchOptions** and it reside).

Some related refactoring is also done to make the eliminating of **FetchStrategy** complete. 

TBH, deleting **FetchStrategy** is not mandatory for sooner or later v5 loading stuff would be deleted and currently there is no need of a FetchOptions implementation class in v6 stuff, but I do think it is a big concept simplification for we only need to work on **FetchOptions** interface alone and we can get rid of the potential confusion between **FetchStrategy** and **FetchOptions**(their relationship, lack of naming convention, potential misuse in method signature or using implementation class instead of interface, etc.).

Let me know whether my rationale of eliminating FetchStrategy is sensible and I am always ready to revert it back.

### CODE REVIEW

File: hibernate-core/src/main/java/org/hibernate/loader/ast/internal/LoaderSelectBuilder.java
Comment (NathanQingyangXu) [inline hibernate-core/src/main/java/org/hibernate/loader/ast/internal/LoaderSelectBuilder.java:719]: This method name is confusing esp. when we have a FetchStrategy class already. I changed it to better reflect its purpose.

File: hibernate-core/src/main/java/org/hibernate/metamodel/mapping/internal/DiscriminatedAssociationAttributeMapping.java
Comment (NathanQingyangXu) [inline hibernate-core/src/main/java/org/hibernate/metamodel/mapping/internal/DiscriminatedAssociationAttributeMapping.java:78]: The above is a perfect example why this PR is necessary.

File: hibernate-core/src/main/java/org/hibernate/sql/results/graph/Fetchable.java
Comment (NathanQingyangXu) [inline hibernate-core/src/main/java/org/hibernate/sql/results/graph/Fetchable.java:22]: I think the above comment could be deleted after this PR is merged.

**Review by beikov — APPROVED** [review]
LGTM but I don't know for sure if this is what @sebersole originally intended, so let's wait for his ok before merging this.

File: hibernate-core/src/main/java/org/hibernate/metamodel/mapping/internal/DiscriminatedAssociationAttributeMapping.java
Comment (sebersole) [reply hibernate-core/src/main/java/org/hibernate/metamodel/mapping/internal/DiscriminatedAssociationAttributeMapping.java:78]: Yes, cases like this are exactly why I created this task

**Review by sebersole — APPROVED** [review]
> LGTM but I don't know for sure if this is what @sebersole originally intended, so let's wait for his ok before merging this.

This is more-or-less exactly what I intended.  Relying on the interface instead of the class allows us to save instantiations

**Pull request merged**
